File: CalculatorFX/src/sample/Calculator/Context.java
package sample.Calculator;

public class Context {
    int x;
    int y;
    char operation;
    State state;

    public int getY() {
        return y;
    }

    public int getX() {
        return x;
    }

    public State getState() {
        return state;
    }

    public char getOperation() {
        return operation;
    }

    public Context(){
        state = new StateX();
        state.clear(this);
    }

    public void press (char key){
        switch (key){
            case 'c':
            case 'C': state.clear(this); break;
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9': state.digit(this, key); break;
            case '-':
            case '+':
            case '/':
            case '%':
            case 'p':
            case '*': state.arifm(this, key); break;
            case '=': state.equal(this); break;
            case '<': state.back(this);
        }
    }

    @Override
    public String toString(){
       return "x = "+x+
               " y = "+y+
               " op = "+operation+
               " state = "+state.getClass().getName();
    }
}

File: Calculator/context.py
from calculator import states


class Context:
    def __init__(self):
        self._state = states.StateX()
        self._state.clear(self)
        self._x = 0
        self._y = 0
        self._oper = '+'

    def calc(self, expr):
        for i in list(expr):
            self.press(i)
        return self._x

    def press(self, key):
        if key == 'c':
            self._state.clear(self)
        elif key.isdigit():
            self._state.digit(self, key)
        elif key == '/' or key == '-' or key == '+' or key == '*' or key == '%':
            self._state.arifm(self, key)
        elif key == '=':
            self._state.equal(self)

    @property
    def state(self):
        return self._state

    @state.setter
    def state(self, state):
        self._state = state

    @property
    def x(self):
        return self._x

    @x.setter
    def x(self, x):
        self._x = x

    @property
    def y(self):
        return self._y

    @y.setter
    def y(self, y):
        self._y = y

    @property
    def operation(self):
        return self._oper

    @operation.setter
    def operation(self, operation):
        self._oper = operation

    def __str__(self):
        return 'x={0}, y={1}, op={2}, s={3}'. \
            format(self._x, self._y, self._oper, self._state)
